File: kotlin/known-values/src/test/kotlin/com/blockchaincommons/knownvalues/DirectoryLoaderTest.kt
package com.blockchaincommons.knownvalues

import java.nio.file.Path
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

class DirectoryLoaderTest {
    @Test
    fun testParseRegistryJson() {
        val json =
            """{
                "ontology": {"name": "test"},
                "entries": [
                    {"codepoint": 9999, "name": "testValue", "type": "property"}
                ],
                "statistics": {}
            }""".trimIndent()

        val registry = RegistryFile.fromJson(json)
        assertEquals(1, registry.entries.size)
        assertEquals(9999uL, registry.entries[0].codepoint)
        assertEquals("testValue", registry.entries[0].name)
    }

    @Test
    fun testParseMinimalRegistry() {
        val json = """{"entries": [{"codepoint": 1, "name": "minimal"}]}"""

        val registry = RegistryFile.fromJson(json)
        assertEquals(1, registry.entries.size)
        assertEquals(1uL, registry.entries[0].codepoint)
    }

    @Test
    fun testParseFullEntry() {
        val json =
            """{
                "entries": [{
                    "codepoint": 100,
                    "name": "fullEntry",
                    "type": "class",
                    "uri": "https://example.com/vocab#fullEntry",
                    "description": "A complete entry with all fields"
                }]
            }""".trimIndent()

        val registry = RegistryFile.fromJson(json)
        val entry = registry.entries[0]
        assertEquals(100uL, entry.codepoint)
        assertEquals("fullEntry", entry.name)
        assertEquals("class", entry.entryType)
        assertEquals("https://example.com/vocab#fullEntry", entry.uri)
        assertNotNull(entry.description)
    }

    @Test
    fun testDirectoryConfigDefault() {
        val config = DirectoryConfig.defaultOnly()
        assertEquals(1, config.paths().size)
        assertTrue(config.paths()[0].toString().endsWith(".known-values"))
    }

    @Test
    fun testDirectoryConfigCustomPaths() {
        val config = DirectoryConfig.withPaths(
            listOf(
                Path.of("/a"),
                Path.of("/b"),
            ),
        )
        assertEquals(2, config.paths().size)
        assertEquals(Path.of("/a"), config.paths()[0])
        assertEquals(Path.of("/b"), config.paths()[1])
    }

    @Test
    fun testDirectoryConfigWithDefault() {
        val config = DirectoryConfig.withPathsAndDefault(listOf(Path.of("/custom")))
        assertEquals(2, config.paths().size)
        assertEquals(Path.of("/custom"), config.paths()[0])
        assertTrue(config.paths()[1].toString().endsWith(".known-values"))
    }

    @Test
    fun testLoadFromNonexistentDirectory() {
        val result = loadFromDirectory(Path.of("/nonexistent/path/12345"))
        assertTrue(result.isEmpty())
    }

    @Test
    fun testLoadResultMethods() {
        val result = LoadResult()
        assertEquals(0, result.valuesCount)
        assertFalse(result.hasErrors)

        result.putValue(KnownValue.withName(1uL, "test"))
        assertEquals(1, result.valuesCount)
    }
}

File: typescript/known-values/tests/directory-loader.test.ts
import { expect, test } from 'vitest';
import {
    DirectoryConfig,
    LoadResult,
    loadFromDirectory,
} from '../src/directory-loader.js';
import { KnownValue } from '../src/known-value.js';

test('parse registry JSON', () => {
    const json = `{
        "ontology": {"name": "test"},
        "entries": [
            {"codepoint": 9999, "name": "testValue", "type": "property"}
        ],
        "statistics": {}
    }`;

    const registry = JSON.parse(json);
    expect(registry.entries.length).toBe(1);
    expect(registry.entries[0].codepoint).toBe(9999);
    expect(registry.entries[0].name).toBe('testValue');
});

test('parse minimal registry', () => {
    const json = '{"entries": [{"codepoint": 1, "name": "minimal"}]}';

    const registry = JSON.parse(json);
    expect(registry.entries.length).toBe(1);
    expect(registry.entries[0].codepoint).toBe(1);
});

test('parse full entry', () => {
    const json = `{
        "entries": [{
            "codepoint": 100,
            "name": "fullEntry",
            "type": "class",
            "uri": "https://example.com/vocab#fullEntry",
            "description": "A complete entry with all fields"
        }]
    }`;

    const registry = JSON.parse(json);
    const entry = registry.entries[0];
    expect(entry.codepoint).toBe(100);
    expect(entry.name).toBe('fullEntry');
    expect(entry.type).toBe('class');
    expect(entry.uri).toBe('https://example.com/vocab#fullEntry');
    expect(entry.description).toBeDefined();
});

test('directory config default', () => {
    const config = DirectoryConfig.defaultOnly();
    expect(config.paths.length).toBe(1);
    expect(config.paths[0]).toContain('.known-values');
});

test('directory config custom paths', () => {
    const config = DirectoryConfig.withPaths(['/a', '/b']);
    expect(config.paths.length).toBe(2);
    expect(config.paths[0]).toBe('/a');
    expect(config.paths[1]).toBe('/b');
});

test('directory config with default', () => {
    const config = DirectoryConfig.withPathsAndDefault(['/custom']);
    expect(config.paths.length).toBe(2);
    expect(config.paths[0]).toBe('/custom');
    expect(config.paths[1]).toContain('.known-values');
});

test('load from nonexistent directory', () => {
    const result = loadFromDirectory('/nonexistent/path/12345');
    expect(result).toEqual([]);
});

test('load result methods', () => {
    const result = new LoadResult();
    expect(result.valuesCount).toBe(0);
    expect(result.hasErrors).toBe(false);

    result.values.set(1n, KnownValue.withName(1, 'test'));
    expect(result.valuesCount).toBe(1);
});
